### DISCUSSION TITLE
Using new DJ5.1 LoginRequiredMiddleware with login_not_required erroneously redirects API call to login-page

### DISCUSSION THREAD

**Discussion by RobKuipers** [encode/django-rest-framework#9503] [Potential Issue]
Using the new Django 5.1 `LoginRequiredMiddleware`, I ran into an issue with calling an API that uses Basic Authentication. It suddenly kept redirecting to the login-page.
So I marked my API call with `@login_not_required`, only to find the problem remained!

Moving the `@login_not_required` decorator to the top (setting it as the first decorator, before `@api_view(['GET'])`), circumvented the problem.
But this seems very ugly. With the need to set the `@login_not_required` decorator on all API's, and more importantly, having to know it needs to be first.

I'm of the opinion API-calls should never redirect to a login-page, but rather return the appropriate http-statuscode.
But, where (and how) to implement this? 

Very interested in your opinions/solutions.

**Comment by thomaslrg** [comment]
I'm not sure this is the best solution, but you can create a custom middleware that bypasses `LoginRequiredMiddleware` for specific URL patterns using regex. [django-stronghold](https://github.com/mgrouchy/django-stronghold) had the same type of functionality.


```python
# app/middleware.py

from django.conf import settings
from django.contrib.auth.middleware import LoginRequiredMiddleware
from django.utils.deprecation import MiddlewareMixin
import re


class CustomLoginRequiredMiddleware(LoginRequiredMiddleware):
    def __init__(self, get_response=None):
        self.get_response = get_response
        self.open_urls = [re.compile(url) for url in settings.OPEN_URLS]
        super().__init__(get_response)

    def process_view(self, request, view_func, view_args, view_kwargs):
        for url in self.open_urls:
            if url.match(request.path):
                return None  # Pass through, no login required
        return super().process_view(request, view_func, view_args, view_kwargs)
```

In `settings.py`, define `OPEN_URLS` and use the custom middleware:

```python
# app/settings.py

MIDDLEWARE = [
    # ...
    "app.middleware.CustomLoginRequiredMiddleware",
]

# Regex patterns for paths that bypass LoginRequiredMiddleware
OPEN_URLS = [
    r"^/my-api/.*",
    # ...
]
```

I'm also very interested in opinions and solutions.

**Comment by browniebroke** [comment]
As far as I understand, it's because Django [`LoginRequiredMiddleware` kicks in](https://github.com/django/django/blob/ff308a06047cd60806d604a7cf612e5656ee2ac9/django/contrib/auth/middleware.py#L50-L57) before the DRF view-level auth happens, so as far as Django is concerned, the HTTP request isn't authenticated until the DRF logic kicks in, which AFAIU is done at the view level:

https://github.com/encode/django-rest-framework/blob/337ba211e82628c0a0bfc7340c45b22e725f8162/rest_framework/views.py#L385-L397

https://github.com/encode/django-rest-framework/blob/8e304e1adbb0f99f91a15ed3abd379104bba3b89/rest_framework/request.py#L378-L395

> calling an API that uses Basic Authentication

I would be curious if you see the same behaviour with session auth. My expectation is that it would work, because this relies on a Django built-in auth mechanism, while the others (Basic and token based auth) are DRF specific.

With regards to solutions, one might be for DRF to provide a specialized version of Django's `LoginRequiredMiddleware`. AFAIK, that fits into the maintenance policy to keep up to date with newer Django features.

**Comment by browniebroke** [comment]
I've dug this a bit more to attempt to add compatibility to DRF as part of https://github.com/encode/django-rest-framework/pull/9514 and just realised that DRF already offers a way to make sure all endpoints are authenticated, via the [`DEFAULT_PERMISSION_CLASSES`](https://www.django-rest-framework.org/api-guide/settings/#default_permission_classes) setting.

**Comment by apetryla** [comment]
@browniebroke Came accross this, since I was thinking weather I should be using DRF `DEFAULT_PERMISSION_CLASSES` or a django middleware (concerning login-required functionality; my custom authenticated user checks, such as email-verified, policy-agreed etc.; and specific permission checks for individual views, e.g. roles). And I noticed this in the docs:
 [Note: when you set new permission classes via the class attribute or decorators you're telling the view to ignore the default list set in the settings.py file.](https://www.django-rest-framework.org/api-guide/permissions/#setting-the-permission-policy)
 Which if being the case, doesn't satisfy my use case, since I wanna stack permission checks (login-required + authenticated-related-almost-ubiquitous-checks + individual-view-checks), not overwrite them. And specifying all of them at view-level results in a lot of duplication and is error-prone. At first glance a quick solution to this might be to subclass `BasePermission`, but I felt hesitant going this way as middleware approach seemed cleaner. Please fix me if I misunderstood smth, but it looks like for my use case I'll have to work at django middleware level and `LoginRequiredMiddleware` ain't gonna help much. :)
